addons/procurement/procurement.py

```diff
@@ -60,184 +51,142 @@ class procurement_rule(osv.osv):
-        return [('state', '=', 'exception')] 
-        
-    def unlink(self, cr, uid, ids, context=None):
-        procurements = self.read(cr, uid, ids, ['state'], context=context)
-        unlink_ids = []
-        for s in procurements:
-            if s['state'] == 'cancel':
-                unlink_ids.append(s['id'])
-            else:
-                raise UserError(_('Cannot delete Procurement Order(s) which are in %s state.') % s['state'])
-        return osv.osv.unlink(self, cr, uid, unlink_ids, context=context)
-
-    def create(self, cr, uid, vals, context=None):
-        context = context or {}
-        procurement_id = super(procurement_order, self).create(cr, uid, vals, context=context)
-        if not context.get('procurement_autorun_defer'):
-            self.run(cr, uid, [procurement_id], context=context)
-        return procurement_id
-
-    def do_view_procurements(self, cr, uid, ids, context=None):
+
+    name = fields.Text(string='Description', required=True)
+    origin = fields.Char(string='Source Document', help="Reference of the document that created this Procurement. This is automatically completed by Odoo.")
+    company_id = fields.Many2one('res.company', string='Company', required=True, default=lambda self: self.env.user.company_id)
+    # These two fields are used for shceduling
+    priority = fields.Selection(PROCUREMENT_PRIORITIES, required=True, index=True, track_visibility='onchange', default='1')
+    date_planned = fields.Datetime(string='Scheduled Date', required=True, default=fields.Datetime.now(), index=True, track_visibility='onchange')
+    group_id = fields.Many2one('procurement.group', string='Procurement Group')
+    rule_id = fields.Many2one('procurement.rule', string='Rule', track_visibility='onchange', help="Chosen rule for the procurement resolution. Usually chosen by the system but can be manually set by the procurement manager to force an unusual behavior.")
+    product_id = fields.Many2one('product.product', string='Product', required=True, states={'confirmed': [('readonly', False)]}, readonly=True)
+    product_qty = fields.Float(string='Quantity', digits_compute=dp.get_precision('Product Unit of Measure'), required=True, states={'confirmed': [('readonly', False)]}, readonly=True)
+    product_uom = fields.Many2one('product.uom', string='Product Unit of Measure', required=True, states={'confirmed': [('readonly', False)]}, readonly=True)
+    state = fields.Selection([('cancel', 'Cancelled'), ('confirmed', 'Confirmed'), ('exception', 'Exception'), ('running', 'Running'), ('done', 'Done')], string='Status', required=True, track_visibility='onchange', copy=False, default='confirmed')
+
+    @api.model
+    def _needaction_domain_get(self):
+        return [('state', '=', 'exception')]
+
+    @api.multi
+    def unlink(self):
+        procurements = self.filtered(lambda procurements: procurements.state == 'cancel')
+        for s in (self-procurements):
+            if s.state != 'cancel':
```

Comment:
Here you get all the records which are not in **cancel** state. So, no need to check the condition.